### PR TITLE
add timesync_chrony_custom_settings variable for free-form local configs

### PR DESCRIPTION
**Pull request by mcaubet**
Added custom settings for ```chrony.conf```

### CODE REVIEW

File: tests/tests_ntp.yml [hunk 1]
```diff
@@ -1,6 +1,9 @@
 - name: Configure time synchronization with NTP servers
   hosts: all
   vars:
+    timesync_chrony_custom_settings:
```
Comment (richm) [inline tests/tests_ntp.yml:4]: will need to document this parameter in README.md
will need some sort of test to verify the settings have been applied

**Comment by richm** [comment]
[citest commit:68e955b5f7499ec9ed605f6b86f23bbc5289aa45]

File: README.md [hunk 2]
```diff
@@ -124,3 +124,22 @@ synchronization:
   roles:
     - linux-system-roles.timesync
 ```
+
+Install and configure chrony with multiple NTP servers and custom advanced 
+settings: log ```measurements```,```statistics``` and ```tracking```
+into ```/var/log/chrony```:
```
Comment (richm) [inline README.md:130]: Please use the single backtick for inline text that should be preformatted.  You only need the 3 backtick form for multiline code/preformatted text.

File: README.md [hunk 3]
```diff
@@ -124,3 +124,22 @@ synchronization:
...
+
+
+```yaml
+- hosts: targets
+  vars:
+    timesync_ntp_servers:
+      - hostname: foo.example.com
+      - hostname: bar.example.com
+      - hostname: baz.example.com
+    timesync_chrony_custom_settings:
```
Comment (richm) [inline README.md:140]: please document `timesync_chrony_custom_settings` in the README.md in the Role Variables section - https://github.com/linux-system-roles/timesync/blob/ccc910ca57e34c6b79ca455e1c33c32df92f3bd9/README.md#role-variables

**Comment by richm** [comment]
@mlichvar @pcahyna This seems like a good feature - just wondering if this is the preferred way to handle provider specific settings.

**Comment by richm** [comment]
I had some questions about the ntp settings - I think they apply to the chrony settings as well - see https://github.com/linux-system-roles/timesync/pull/54/files#r469487750 and https://github.com/linux-system-roles/timesync/pull/54/files#r469489022

**Comment by richm** [comment]
[citest]

**Comment by richm** [comment]
travis ci is failing due to yamllint error:
```
yamllint run-test: commands[0] | yamllint .

./tests/tests_ntp.yml

  35:16     error    trailing spaces  (trailing-spaces)
```
you can use `tox -e yamllint` locally to debug this error.  You can use `tox -e molecule` to also look for ansible lint and other issues.

**Comment by mcaubet** [comment]
> I had some questions about the ntp settings - I think they apply to the chrony settings as well - see https://github.com/linux-system-roles/timesync/pull/54/files#r469487750 and https://github.com/linux-system-roles/timesync/pull/54/files#r469489022

If user wants to remove a specific setting he just needs to remove it from the list. Custom settings will literally apply any string in the list. In example, something like:

```yaml
timesync_chrony_custom_settings:
  - "initstepslew 60 pstime1.psi.ch pstime2.psi.ch pstime3.psi.ch"
  - "logdir /var/log/chrony"
  - "log measurements statistics tracking"
```
will add the above lines at the end of chrony.conf. If we want to remove the logging part:
```yaml
timesync_chrony_custom_settings:
  - "initstepslew 60 pstime1.psi.ch pstime2.psi.ch pstime3.psi.ch"
```
This will remove the logdir and log lines. Same if you just want to apply a change to these lines, as soon as one updates the line new line will apply. Same rule will apply here: https://github.com/linux-system-roles/timesync/pull/54 I just wrote something very similar in order to keep coherency. The difference is that on my pull request custom settings are added to the end of the file, will on #54 are added to the top (I really prefer to keep predefined settings on top and custom settings at the end, as it can be a random list of settings). The only problem is that no comments will be added, unless one tries to do something like:

```yaml
timesync_chrony_custom_settings:
  - "# Allow chronyd to make a rapid measurement of the system and correct clock error at boot time"
  - "initstepslew 60 pstime1.psi.ch pstime2.psi.ch pstime3.psi.ch"
  - "# Select which information is logged."
  - "logdir /var/log/chrony"
  - "log measurements statistics tracking"
```
The way for adding comments can be improved by enforcing a comment for each new line of settings (it is, instead of a list of strings, a list of hashes with 2 string elements), but in general terms it would really make code more complex and more difficult to understand. In general, predefined settings are in general enough, but on specific use cases one needs to use advanced settings (this is our case for instance, we are managing few datacenters as well user workstations in a complex network setup, so we need to have tools for debugging and optimizing time synchronization).

**Comment by mcaubet** [comment]
I had to create a specific test `tests_chrony.yml` for chrony and leave `tests_ntp.yaml` as it was in the master branch: looks like for RHEL6-like systems the default provider is NTP, so `tests_ntp.yaml` with custom settings was failing with the extended chrony settings (as these systems were using NTP by default for this test).

**Comment by richm** [comment]
> If user wants to remove a specific setting he just needs to remove it from the list. 

There is no support for incremental settings.  The list defined in `timesync_chrony_custom_settings` will replace whatever already exists in the config file.  The `template` module is idempotent if you run the playbook again with the exact same `timesync_chrony_custom_settings`.  As far as comments go, I think adding them to `timesync_chrony_custom_settings` is fine - I don't think any custom handling for comments is necessary.

How to remove custom settings?  For example, if I have already run the role with
```yaml
timesync_chrony_custom_settings:
  - "initstepslew 60 time1.domain.ch time2.domain.ch time3.domain.ch"
  - "logdir /var/log/chrony"
  - "log measurements statistics tracking"
```
Now I want to remove all of these settings.  I suppose one way would be to do something like this:
```yaml
timesync_chrony_custom_settings:
  - "# no custom settings"
```
which isn't great, but it works.  Any suggestions?

**Comment by mcaubet** [comment]
Hi,

If one wants to remove all settings, by do not defining
timesync_chrony_custom_settings would be enough. It is, if variable is not
defined or list length is 0, should write nothing in the config file.
Setting is an empty list by default.

**Comment by richm** [comment]
> Hi, If one wants to remove all settings, by do not defining timesync_chrony_custom_settings would be enough. It is, if variable is not defined or list length is 0, should write nothing in the config file. Setting is an empty list by default.

OK.  Right, because the entire file would be created from scratch every time the role is run.

**Comment by mcaubet** [comment]
> > Hi, If one wants to remove all settings, by do not defining timesync_chrony_custom_settings would be enough. It is, if variable is not defined or list length is 0, should write nothing in the config file. Setting is an empty list by default.
> 
> OK. Right, because the entire file would be created from scratch every time the role is run.

Hi,

below there is an example from switching from this variable configuration:

```yaml
# linux-system-roles.timesync settings
timesync_chrony_custom_settings:
  - "# Allow chronyd to make a rapid measurement of the system and correct clock error at boot time"
  - "initstepslew 60 pstime1.psi.ch pstime2.psi.ch pstime3.psi.ch"
  - "# Select which information is logged."
  - "logdir /var/log/chrony"
  - "log measurements statistics tracking"
timesync_ntp_provider: chrony
timesync_ntp_servers:
  - hostname: pstime1.psi.ch
  - hostname: pstime2.psi.ch
  - hostname: pstime3.psi.ch
```

to this configuration (removing custom settings):

```yaml
# linux-system-roles.timesync settings
# timesync_chrony_custom_settings:
#   - "# Allow chronyd to make a rapid measurement of the system and correct clock error at boot time"
#   - "initstepslew 60 pstime1.psi.ch pstime2.psi.ch pstime3.psi.ch"
#   - "# Select which information is logged."
#   - "logdir /var/log/chrony"
#   - "log measurements statistics tracking"
timesync_ntp_provider: chrony
timesync_ntp_servers:
  - hostname: pstime1.psi.ch
  - hostname: pstime2.psi.ch
  - hostname: pstime3.psi.ch
```

Is just a diff and looks like is properly removing only that part of the config file. I guess is what you expected, is it correct?

```shell
TASK [timesync : Generate chrony.conf file] *******************************************************************************************************************************************************************************************************************************************************************************************************************************************
--- before: /etc/chrony.conf
+++ after: /root/.ansible/tmp/ansible-local-11683pw2yr63/tmpmy0upwh_/chrony.conf.j2
@@ -13,9 +13,3 @@
 # Record the rate at which the system clock gains/losses time.
 driftfile /var/lib/chrony/drift
 
-# Custom settings
-# Allow chronyd to make a rapid measurement of the system and correct clock error at boot time
-initstepslew 60 pstime1.psi.ch pstime2.psi.ch pstime3.psi.ch
-# Select which information is logged.
-logdir /var/log/chrony
-log measurements statistics tracking

changed: [rhel-8-dev-7a95e9bb.psi.ch]
```
Otherwise let me know.

**Comment by richm** [comment]
Thanks.  Looks good.

**Comment by richm** [comment]
[citest]

**Comment by richm** [comment]
@mlichvar ok to merge?

**Comment by mlichvar** [comment]
Looks good to me. Thanks.

**Comment by mlichvar** [comment]
But please squash the commits before merge.

**Comment by richm** [comment]
@mcaubet please squash the commits into a single commit

**Comment by richm** [comment]
looks like you squashed your commits into a merge commit?  try this instead: https://github.com/wprig/wprig/wiki/How-to-squash-commits

**Comment by mcaubet** [comment]
> @mcaubet looks like you squashed your commits into a merge commit? try this instead: https://github.com/wprig/wprig/wiki/How-to-squash-commits

Hi, I just resolved some conflicts. I will try to locally squash it (I already did, but my master branch is not in sync with yours). However, I thought squashing commits + merging was possible when merging a pull request, isn't it?

**Comment by mcaubet** [comment]
There still was a conflict that I had to resolve. Is it ok now?

**Comment by richm** [comment]
> There still was a conflict that I had to resolve. Is it ok now?

You still have two commits here.  And one is a merge commit.  Did you follow the procedure listed at https://github.com/wprig/wprig/wiki/How-to-squash-commits ?

**Comment by richm** [comment]
@mcaubet the commit code looks good - however, the commit message contains everything except the description of `timesync_chrony_custom_settings`

**Comment by richm** [comment]
@mcaubet you can use `git commit --amend` to edit the commit message, then `git push -f mcaubet timesync_chrony_custom_settings` to update the PR with the commit with the amended message

**Comment by mcaubet** [comment]
> @mcaubet you can use `git commit --amend` to edit the commit message, then `git push -f mcaubet timesync_chrony_custom_settings` to update the PR with the commit with the amended message

@richm sorry, I updated the message. Yes, I am not very used to these formalities, but I completely agree with you: is important to keep it right. Thanks for pointing this out. I hope now is fine otherwise let me know.

Best regards and thanks for your help,
Marc

**Comment by richm** [comment]
[citest commit:28a71a3ecdeb157501747b5d475ab38b7025609b]

**Pull request merged**